File: backend/models/market.py
from pydantic import BaseModel
from typing import Optional, List
from enum import Enum

class OilType(str, Enum):
    WTI = "WTI"
    BRENT = "BRENT"

class OHLCVBar(BaseModel):
    time: int
    open: float
    high: float
    low: float
    close: float
    volume: float 

class Quote(BaseModel):
    symbol: str
    name: str
    price: float
    change: float
    change_percent: float
    volume: float
    currency: str
    delay_minutes: int

class MarketStat(BaseModel):
    symbol: str
    market_cap: Optional[float] = None
    pe_ratio: Optional[float] = None
    week_high_52: Optional[float] = None
    week_low_52: Optional[float] = None
    dividend_yield: Optional[float] = None
    avg_volume: Optional[float] = None
    
class MarketHistoryResponse(BaseModel):
    symbol: str
    timeframe: str
    bars: List[OHLCVBar]
    delay_minutes: int

class IndexItem(BaseModel):
    symbol: str
    name: str
    price: float
    change_percent: float
    region: str
    currency: str

class MetalQuote(BaseModel):
    symbol: str
    name: str
    spot_price: float
    futures_price: Optional[float] = None
    currency: str
    change: float
    change_percent: float
    unit: str
    delay_minutes: int

class CommodityQuote(BaseModel):
    symbol: str
    name: str
    price: float
    change: float
    change_percent: float
    currency: str
    unit: str
    category: str
    delay_minutes: int

File: frontend/src/types/market.ts
export interface OHLCVBar {
    time: number
    open: number
    high: number
    low: number
    close: number
    volume: number
}

export interface Quote {
    symbol: string
    name: string
    price: number
    change: number
    change_percent: number
    volume: number
    currency: string
    exchange: string
    delay_minutes: number
}

export interface MarketStat {
    symbol: string
    market_cap: string
    pe_ratio: number | null
    week_high_52: number | null
    week_low_52: number | null
    dividend_yield: number | null
    avg_volume: number | null
}

export interface MarketHistoryResponse {
    symbol: string
    timeframe: string
    bars: OHLCVBar[]
    delay_minutes: number
}

export type DataFreshness =
| 'realtime'
| 'near-realtime'
| 'delayed'
| 'end-of-day'

export function getDataFreshness(delayMinutes: number): DataFreshness {
    if(delayMinutes === 0) return 'realtime'
    if(delayMinutes <= 5) return 'near-realtime'
    if(delayMinutes <= 60) return 'delayed'
    return 'end-of-day'
}